core/src/validator.rs

```diff
@@ -1925,85 +1929,106 @@ fn maybe_warp_slot(
     Ok(())
 }
 
+/// Searches the blockstore for data shreds with the incorrect shred version.
 fn blockstore_contains_bad_shred_version(
     blockstore: &Blockstore,
     start_slot: Slot,
-    shred_version: u16,
-) -> bool {
-    let now = Instant::now();
+    expected_shred_version: u16,
+) -> Result<bool, BlockstoreError> {
+    const TIMEOUT_MS: u64 = 60000;
+    let timer = AtomicInterval::default();
     // Search for shreds with incompatible version in blockstore
-    if let Ok(slot_meta_iterator) = blockstore.slot_meta_iterator(start_slot) {
-        info!("Searching for incorrect shreds..");
-        for (slot, _meta) in slot_meta_iterator {
-            if let Ok(shreds) = blockstore.get_data_shreds_for_slot(slot, 0) {
-                for shred in &shreds {
-                    if shred.version() != shred_version {
-                        return true;
-                    }
-                }
-            }
-            if now.elapsed().as_secs() > 60 {
-                info!("Didn't find incorrect shreds after 60 seconds, aborting");
-                return false;
+    let slot_meta_iterator = blockstore.slot_meta_iterator(start_slot)?;
+
+    info!("Searching blockstore for shreds with incorrect version..");
+    for (slot, _meta) in slot_meta_iterator {
+        let shreds = blockstore.get_data_shreds_for_slot(slot, 0)?;
+        for shred in &shreds {
+            if shred.version() != expected_shred_version {
+                return Ok(true);
             }
         }
+        if timer.elapsed_ms() > TIMEOUT_MS {
+            info!("Didn't find incorrect shreds after 60 seconds, aborting");
+            break;
+        }
     }
-    false
+    Ok(false)
 }
 
+/// If the blockstore contains any shreds with the incorrect shred version,
+/// copy them to a backup blockstore and purge them from the actual blockstore.
 fn backup_and_clear_blockstore(
     ledger_path: &Path,
     config: &ValidatorConfig,
     start_slot: Slot,
-    shred_version: u16,
-) {
+    expected_shred_version: u16,
+) -> Result<(), BlockstoreError> {
     let blockstore =
-        Blockstore::open_with_options(ledger_path, blockstore_options_from_config(config)).unwrap();
+        Blockstore::open_with_options(ledger_path, blockstore_options_from_config(config))?;
     let do_copy_and_clear =
-        blockstore_contains_bad_shred_version(&blockstore, start_slot, shred_version);
+        blockstore_contains_bad_shred_version(&blockstore, start_slot, expected_shred_version)?;
+    if !do_copy_and_clear {
```

Comment:
prolly prefer the original nesting on (un)inverted conditional to the early return for maintainability's sake

Reply:
it's more about future maintainability and self-defense against bug introduction. way harder to do when every logical condition is scoped